e2e_tests/tests/cluster/test_ship_logs.py

```diff
@@ -0,0 +1,512 @@
+            )
+            with ShipLogServer() as srv:
+                exit_code = self.run_ship_logs(srv.master_url(), cmd)
+            assert exit_code == 0, exit_code
+            if "".join(srv.logs) == "1\n2\n3\n4\n5\n6\n":
+                # Success
+                break
+            elif timeout == timeouts[-1]:
+                # Failed, even on the highest timeout
+                raise ValueError("".join(srv.logs))
+
+    @pytest.mark.e2e_cpu
+    def test_signal_forwarding(self) -> None:
+        cmd = mkcmd(
+            """
+            import signal
+            import time
+
+            def handle_sigint(*arg):
+                print("caught sigint!")
+
+            signal.signal(signal.SIGINT, handle_sigint)
+
+            print("ready!", flush=True)
+
+            time.sleep(5)
+            """
+        )
+        with ShipLogServer() as srv:
+            # Start a subprocess so we can signal it.
+            env = {
+                "DET_MASTER": srv.master_url(),
+                "DET_SESSION_TOKEN": "token",
+                "DET_TASK_LOGGING_METADATA": "{}",
+                "DET_SHIPPER_EMIT_STDOUT_LOGS": "1",
+            }
+            fullcmd = [sys.executable, "-u", ship_logs.__file__] + cmd
+            p = subprocess.Popen(fullcmd, env=env, stdout=subprocess.PIPE)
+            assert p.stdout
+            try:
+                # Wait for the granchild log to indicate signals are set up.
+                for line in p.stdout:
+                    if b"ready!" in line:
+                        break
+                # Send a signal that is caught and logged, to test signal forwarding.
+                p.send_signal(signal.SIGINT)
+                for line in p.stdout:
+                    if b"caught sigint!" in line:
+                        break
+                # Send a signal that is not caught, to test for signal exit codes.
+                p.send_signal(signal.SIGTERM)
+                exit_code = p.wait()
+            finally:
+                p.kill()
+                p.wait()
+        assert exit_code == 128 + signal.SIGTERM, exit_code
+        assert "".join(srv.logs) == "ready!\ncaught sigint!\n", srv.logs
+
+    @pytest.mark.e2e_cpu
+    def test_exit_wait_time(self) -> None:
+        cmd = mkcmd("print('hello world')")
+        # We need a misbehaving server to guarantee the shipper times out.
+        # This misbehaving server will listen without ever accepting.
+        with socket.socket() as listener:
+            listener.bind(("127.0.0.1", 0))
+            listener.listen(10)
+            _, port = listener.getsockname()
+            master_url = f"http://127.0.0.1:{port}"
+            start = time.time()
+            exit_code = self.run_ship_logs(master_url, cmd, log_wait_time=0.1)
+            end = time.time()
+            assert exit_code == 0, exit_code
+            assert end - start < 1, end - start
+
+    @pytest.mark.e2e_cpu
+    def test_entrypoint_not_found(self) -> None:
```

Comment:
nit here on names, what's being tested in that the proper exit codes (here and in the next test) are returned on a failure. If that was in the name, it'd be more apparent.